Sources/Subs-Graphics.php

```diff
@@ -306,10 +302,69 @@ function checkGD()
 	return true;
 }
 
+function checkImagick()
+{
+	return class_exists('Imagick');
+}
+
 function resizeImageFile($source, $destination, $max_width, $max_height, $preferred_format = 0)
 {
 	global $sourcedir;
+	@ini_set('memory_limit', '90M');
+	
+	//Can we do this with ImageMagick?
+	if(checkImagick())
+	{
+		// Get the image file, we have to work with something after all
+		if(!copy($source, $destination))
+			return false;
+		try
+		{
+			$imagick = new Imagick();
+			$imagick->readImage($destination);
+			$imageformat = $imagick->getImageFormat();
+			$gif = ($imageformat === 'GIF');
+			if ($gif) // If this is a GIF, we need to coalesce all frames to ensure they're all of the same size
+				$imagick = $imagick->coalesceImages();
+			do
+			{
+				$imagick->gammaImage(0.45454545); // Work around gamma error, see http://www.ericbrasseur.org/gamma.html
+
+				// if either the width or height are null, set the current dimension to be the maximum
+				if($max_width === null)
+					$max_width = $imagick->getImageWidth();
+				if($max_height === null)
+					$max_height = $imagick->getImageHeight();
+				
+				// FILTER_CATROM has been chosen as a middle ground between performance and quality. Could be changed to FILTER_BOX if performance is struggling
+				$imagick->resizeImage($max_width, $max_height, Imagick::FILTER_CATROM, 1, true, false);
+				$imagick->gammaImage(2.2); // Work around gamma error, see http://www.ericbrasseur.org/gamma.html
+			} while ($imagick->nextImage());
+
+			if ($gif) // Now that all frames have been resized, we can deconstruct it back to a progressive GIF
+				$imagick = $imagick->deconstructImages();
+
+			//If the preferred format is set to 3 (PNG), respect it for things that aren't gifs
+			if(!$gif && $preferred_format == 3)
```

Comment:
Does the Imagick library provide a constant we can use instead of a magic number here?

Reply:
This magic number comes from SMF's internals, presumably inspired by GD. It could certainly be changed or omitted entirely, but it would involve changing things other than just image handling (e.g. how these options are represented within SMF internally). And then we'd have to offload the magic numbers to legacy functions anyway.